include/os/windows/spl/sys/kmem_impl.h

```diff
@@ -0,0 +1,501 @@
+	uint64_t arc_no_grow;
+
+	/*
+	 * Cache properties
+	 */
+	char		cache_name[KMEM_CACHE_NAMELEN + 1];
+	uint32_t	cache_bufsize;	/* object size */
+	uint32_t	cache_align;	/* object alignment */
+	int		(*cache_constructor)(void *, void *, int);
+	void		(*cache_destructor)(void *, void *);
+	void		(*cache_reclaim)(void *);
+	kmem_cbrc_t	(*cache_move)(void *, void *, uint32_t, void *);
+	void		*cache_private;	/* opaque arg to callbacks */
+	vmem_t		*cache_arena;	/* vmem source for slabs */
+	int		cache_cflags;	/* cache creation flags */
+	int		cache_flags;	/* various cache state info */
+	uint32_t	cache_mtbf;	/* induced alloc failure rate */
+	uint32_t	cache_pad1;	/* compiler padding */
+	kstat_t		*cache_kstat;	/* exported statistics */
+	list_node_t	cache_link;	/* cache linkage */
+
+	/*
+	 * Slab layer
+	 */
+	kmutex_t	cache_lock;		/* protects slab layer */
+
+	uint32_t	cache_chunksize; 	/* buf + alignment [+ debug] */
+	uint32_t	cache_slabsize;		/* size of a slab */
+	uint32_t	cache_maxchunks; 	/* max buffers per slab */
+	uint32_t	cache_bufctl;		/* buf-to-bufctl distance */
+	uint32_t	cache_buftag;		/* buf-to-buftag distance */
+	uint32_t	cache_verify;		/* bytes to verify */
+	uint32_t	cache_contents;		/* bytes of saved content */
+	uint32_t	cache_color;		/* next slab color */
+	uint32_t	cache_mincolor;		/* maximum slab color */
+	uint32_t	cache_maxcolor;		/* maximum slab color */
+	uint32_t	cache_hash_shift;	/* get to interesting bits */
+	uint32_t	cache_hash_mask;	/* hash table mask */
+	list_t		cache_complete_slabs;	/* completely allocated slabs */
+	uint32_t	cache_complete_slab_count;
+	avl_tree_t	cache_partial_slabs;	/* partial slab freelist */
+	uint32_t	cache_partial_binshift;	/* for AVL sort bins */
+	kmem_cache_t	*cache_bufctl_cache;	/* source of bufctls */
+	kmem_bufctl_t	**cache_hash_table;	/* hash table base */
+	kmem_defrag_t	*cache_defrag;		/* slab consolidator fields */
+
+	/*
+	 * Depot layer
+	 */
+	kmutex_t	cache_depot_lock;	/* protects depot */
+	kmem_magtype_t	*cache_magtype;		/* magazine type */
+	kmem_maglist_t	cache_full;		/* full magazines */
+	kmem_maglist_t	cache_empty;		/* empty magazines */
+	void	*cache_dumpfreelist;		/* heap during crash dump */
+	void	*cache_dumplog;			/* log entry during dump */
+
+	/*
+	 * Per CPU structures
+	 */
+	// XNU adjust to suit offsetof
+	kmem_cpu_cache_t    cache_cpu[1];	/* per-cpu data */
+};
+
+typedef struct kmem_cpu_log_header {
+	kmutex_t	clh_lock;
+	char		*clh_current;
+	uint32_t	clh_avail;
+	int		clh_chunk;
+	int		clh_hits;
+#if defined(SPL_DEBUG_MUTEX)
+	char		clh_pad[128 - sizeof (kmutex_t) - sizeof (char *) -
+	    sizeof (uint32_t) - 2 * sizeof (int)];
+#else
+	char		clh_pad[128 - sizeof (kmutex_t) - sizeof (char *) -
+	    sizeof (uint32_t) - 2 * sizeof (int)];
+#endif
+} kmem_cpu_log_header_t;
+
+typedef struct kmem_log_header {
+	kmutex_t	lh_lock;
+	char		*lh_base;
+	int		*lh_free;
+	uint32_t	lh_chunksize;
+	int		lh_nchunks;
+	int		lh_head;
+	int		lh_tail;
+	int		lh_hits;
```

Comment:
Everything that is `int` here probably should be `uint32_t`.

Reply:
Probably true - but hesitant to change anything from original Solaris source, without looking deeper at it :)

Reply:
although I see no reason these should be negative - but i know kmem itself does use negatives in some areas

Reply:
Not here. I checked.